deparse.c

```diff
@@ -682,6 +715,16 @@ sqlite_foreign_expr_walker(Node *node,
 					return false;
 				}
 
+				if (oprleft != InvalidOid && oprright != InvalidOid)
+				{
+					pg_dataTypeName = TypeNameToString(makeTypeNameFromOid(oprleft, -1));
+					if(!is_pg_builtin_Oid(oe->opno) && (
```

Comment:
No, @son-phamngoc. After this condition removing many pushed down to SQLite `WHERE` conditions transformed to PostgreSQL side filter. You are wrong here.

Reply:
At the version that I commented, both of your conditions were checking if `oe->opno` is built-in operator. That's why I commented that no need to check twice for the same condition.  
Now, I see that you changed the condition to check `oprleft`.  
How about `oprright`? I think it should also be checked if it is built-in.  

And one more point here, could you fix the last parenthesis? Following Postgres style, it should not be at a new line.
```
if (!sqlite_is_builtin(oprleft) && !sqlite_is_builtin(oprright)
   (!listed_datatype_oid(oprleft, -1, postGisSQLiteCompatibleTypes) ||
    oprleft != oprright))
```

Reply:
> How about `oprright`? I think it should also be checked if it is built-in.

Thanks, @son-phamngoc ! Please check my final variant here and description:
`!sqlite_is_builtin(oprleft)` not standard left operand or `!sqlite_is_builtin(oprright))` not standard right operand if this are different and not PostGIS data types will not pushed down.

Reply:
Could you explain why you only check `!listed_datatype_oid(oprleft, -1, postGisSQLiteCompatibleTypes)` but does not check `!listed_datatype_oid(oprright, -1, postGisSQLiteCompatibleTypes)`?

Reply:
@mkgrgis The condition here is quite confused. Could you specify clearly what specification that you are going to implement?  
By looking at this, you are only handle the case that there are both `oprleft` `oprright`. However, there will be the case that there is only `oprleft` or only `oprright`. I think you will miss that case. It is better to treat `oprleft` and `oprright` separately.

Reply:
> Could you specify clearly what specification that you are going to implement?

Before this lines (predicate from 693) we have internal pg operators or `=` operator. Next _we should ensure this is one of PostGIS `=` operators_ and return `false` otherwise. This was my simple idea with bad implementation around of the lines.

Reply:
I have an idea (not debug yet):  
```
bool				non_builtin_equal_opr = false;

/* Change line 693 */
if (!sqlite_is_builtin(oe->opno))
{
	if ((strcmp(cur_opname, "=") != 0))
		non_builtin_equal_opr = true;       /* Set it to true for later check */
	else
		return false;
}

...

/* Replace your code by this code block and also add comment */
if (non_builtin_equal_opr)
{
	/* If left operand is not PostGIS supported data type, not push down */
	if (sqlite_is_builtin(oprleft) || (!listed_datatype_oid(oprleft, -1, postGisSQLiteCompatibleTypes)))
		return false;
					
	/* If right operand is not PostGIS supported data type, not push down */
	if (sqlite_is_builtin(oprright) || (!listed_datatype_oid(oprright, -1, postGisSQLiteCompatibleTypes)))
		return false;

	/* Log operator for potential pushing down */
	elog(DEBUG2, "sqlite_fdw : %s operator", cur_opname);
}
```
Could you try it?

Reply:
Fixed in https://github.com/pgspider/sqlite_fdw/pull/96/commits/baf120faa097ab87c79fac80067525fcbf6da4ce with some more universal naming.

Reply:
Look like there are no big problems now, only failed _inverted_ `--Testcase 39:` of both `nogis` and `postgis` tests. One of my pushing predicates is not correct.

Reply:
PostGIS tests are fixed in https://github.com/pgspider/sqlite_fdw/pull/96/commits/fbe71c9954d3cb5f4e0b12ac5c8d0b934c539ff5. I'll add requested TCs about operators etc. and fix `nogis` test.

Reply:
@mkgrgis Test case 39 in `nogis` shows wrong deparsed constant value and wrong result. Could you check again?  
```
--Testcase 39:
EXPLAIN (VERBOSE, COSTS OFF)
SELECT "i", gm, gg, t FROM "types_PostGIS" WHERE gm = '0101000020e6100000bf72ce99fe763e40ed4960730ed84d40'::geometry;
                                                                                           QUERY PLAN                                                                                           
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Foreign Scan on public."types_PostGIS"
   Output: i, gm, gg, t
   SQLite query: SELECT `i`, `gm`, `gg`, `t` FROM main."types_PostGIS" WHERE ((`gm` = X'3031303130303030323065363130303030306266373263653939666537363365343065643439363037333065643834643430'))
(3 rows)

--Testcase 40: ERR - no GIS data support
SELECT "i", gm, gg, t FROM "types_PostGIS" WHERE gm = '0101000020e6100000bf72ce99fe763e40ed4960730ed84d40'::geometry;
 i | gm | gg | t 
---+----+----+---
(0 rows)
```

Reply:
@son-phamngoc , look like https://github.com/pgspider/sqlite_fdw/pull/96/commits/aab56afbc1d4b0e999dc4b90a3dbf69b58bba0b4 is not helpable against this problem. Is here something about _complementary node_, how do you think?